### ISSUE TITLE
Video Encode API: Encoder gets stuck after varying amount of time.

### ISSUE THREAD

**Issue by Xaymar** [GPUOpen-LibrariesAndSDKs/AMF#19]
Reported by one my users and shortly afterwards reproduced, it seems that encoding will get stuck at some point, never returning from SubmitInput or QueryOutput - it simply doesn't do things anymore. 

### Parameters (Applied in the order shown)
* Quality Preset: Speed
* Usage: Transcoding
* Profile: Main
* Profile Level: 41
* Frame Size: 1280, 720
* Frame Rate: 60, 1
* Rate Control Method: Constant Bitrate
* Rate Control Skip Frame: Disabled (false)
* Minimum QP: 18
* Maximum QP: 51
* Target Bitrate: 3000000
* Peak Bitrate: 3000000
* VBV Buffer Size: 3000000
* Initial VBV Buffer Fullness: 100% (64)
* Filler Data: Enabled (true)
* IDR Period: 120
* B-Pictures Pattern: 3
* B-Reference: Enabled (true)
* B-Picture Delta QP: 0
* Reference B-Picture Delta QP: 0
* Scan Type: Progressive
* Half Pixel Motion Estimation: Enabled (true)
* Quarter Pixel Motion Estimation: Enabled (true)
* Frame Size: 1280, 720
* Frame Rate: 60, 1

### Dump File

* xaymar.com: http://cdn.xaymar.com/private/2016/09/11-09-2016_obs64_2.7z
* mega.co.nz: https://mega.nz/#!eBdxlbhJ!MxAcaFGigTQ2OusIHGocdiQMsD9O0sUO0pgtp4BGhBE

**Comment by Xaymar** [comment]
Occasionally encoding will return to normal and hiccup every few minutes. This is extremely rare.

**Comment by Xaymar** [comment]
Can easily be reproduced by loading the dashboard on Twitch for example - it will cause a short hiccup because the encoding task has a much lower priority than it should. We need the ability to specify priority for the task again!

**Comment by MikhailAMD** [comment]
Hi,
Please try to add InitOpenCL() after InitDX9() and InitDX11(). This will move some operations into Compute HW queue and make it easier on 3D queue.

**Comment by Xaymar** [comment]
Right, so I have the changes that you suggested in #12 (and here) implemented now. Seems to "fix" it, except that it now spams log messages. I've updated the code in my repository with what I have right now - do tell me if it's wrong somewhere.

```
21:17:08.610: [AMFTrace::AMFSurfaceImpl]    Error: ..\..\..\..\..\runtime\src\core\SurfaceImpl.cpp(210):Assertion failed:pInteropDevice != NULL
21:17:08.610: [AMF Encoder] Packet: Type(2), PTS(1855,    30915430), DTS(1850,    30833334), Size(    3107)
21:17:08.626: [AMF Encoder] Packet: Type(3), PTS(1852,    30865432), DTS(1851,    30850001), Size(    2772)
21:17:08.626: [AMFTrace::AMFSurfaceImpl]    Error: ..\..\..\..\..\runtime\src\core\SurfaceImpl.cpp(210):Assertion failed:pInteropDevice != NULL
21:17:08.643: [AMFTrace::AMFSurfaceImpl]    Error: ..\..\..\..\..\runtime\src\core\SurfaceImpl.cpp(210):Assertion failed:pInteropDevice != NULL
21:17:08.643: [AMF Encoder] Packet: Type(3), PTS(1853,    30882098), DTS(1852,    30866668), Size(    2807)
21:17:08.660: [AMF Encoder] Packet: Type(3), PTS(1854,    30898764), DTS(1853,    30883334), Size(    2969)
21:17:08.660: [AMFTrace::AMFSurfaceImpl]    Error: ..\..\..\..\..\runtime\src\core\SurfaceImpl.cpp(210):Assertion failed:pInteropDevice != NULL
21:17:08.676: [AMF Encoder] Packet: Type(2), PTS(1859,    30982094), DTS(1854,    30900001), Size(    2575)
21:17:08.676: [AMFTrace::AMFSurfaceImpl]    Error: ..\..\..\..\..\runtime\src\core\SurfaceImpl.cpp(210):Assertion failed:pInteropDevice != NULL
21:17:08.693: [AMFTrace::AMFSurfaceImpl]    Error: ..\..\..\..\..\runtime\src\core\SurfaceImpl.cpp(210):Assertion failed:pInteropDevice != NULL
21:17:08.693: [AMF Encoder] Packet: Type(3), PTS(1856,    30932096), DTS(1855,    30916668), Size(    2863)
21:17:08.710: [AMF Encoder] Packet: Type(3), PTS(1857,    30948762), DTS(1856,    30933334), Size(    2086)
21:17:08.710: [AMFTrace::AMFSurfaceImpl]    Error: ..\..\..\..\..\runtime\src\core\SurfaceImpl.cpp(210):Assertion failed:pInteropDevice != NULL
21:17:08.727: [AMF Encoder] Packet: Type(3), PTS(1858,    30965428), DTS(1857,    30950001), Size(    1896)
21:17:08.727: [AMFTrace::AMFSurfaceImpl]    Error: ..\..\..\..\..\runtime\src\core\SurfaceImpl.cpp(210):Assertion failed:pInteropDevice != NULL
21:17:08.744: [AMF Encoder] Packet: Type(2), PTS(1863,    31048758), DTS(1858,    30966668), Size(    2581)
21:17:08.744: [AMFTrace::AMFSurfaceImpl]    Error: ..\..\..\..\..\runtime\src\core\SurfaceImpl.cpp(210):Assertion failed:pInteropDevice != NULL
21:17:08.760: [AMF Encoder] Packet: Type(3), PTS(1860,    30998760), DTS(1859,    30983334), Size(    2110) 
21:17:08.760: [AMFTrace::AMFSurfaceImpl]    Error: ..\..\..\..\..\runtime\src\core\SurfaceImpl.cpp(210):Assertion failed:pInteropDevice != NULL
21:17:08.776: [AMFTrace::AMFSurfaceImpl]    Error: ..\..\..\..\..\runtime\src\core\SurfaceImpl.cpp(210):Assertion failed:pInteropDevice != NULL
21:17:08.776: [AMF Encoder] Packet: Type(3), PTS(1861,    31015426), DTS(1860,    31000001), Size(    1855)
21:17:08.793: [AMFTrace::AMFSurfaceImpl]    Error: ..\..\..\..\..\runtime\src\core\SurfaceImpl.cpp(210):Assertion failed:pInteropDevice != NULL
21:17:08.793: [AMF Encoder] Packet: Type(3), PTS(1862,    31032092), DTS(1861,    31016668), Size(    2666)
21:17:08.810: [AMF Encoder] Packet: Type(2), PTS(1867,    31115422), DTS(1862,    31033334), Size(    1817)
21:17:08.810: [AMFTrace::AMFSurfaceImpl]    Error: ..\..\..\..\..\runtime\src\core\SurfaceImpl.cpp(210):Assertion failed:pInteropDevice != NULL
```

**Comment by MikhailAMD** [comment]
Can I see your latest code with this?

Mikhail

**Comment by Xaymar** [comment]
The code is right here: https://github.com/Xaymar/OBS-AMD-Advanced-Media-Framework

**Comment by MikhailAMD** [comment]
Please separate issues. Let's consider this one is about stuck. I reviewed the dump. It doesn't have a thread that sits inside amfrt64.DLL. There are two threads sitting inside the plug-in (enc-amf.dll, thread IDs = 3332, 17724). What is the reason to believe that the plug-in is blocked by AMF?

**Comment by MikhailAMD** [comment]
About hick-ups: By default color space conversion in OBS is done on D3D HW queue. It may interfere with other applications. AMF provides you an option to use a different queue: Compute HW queue for the conversion. In this case there is much less interference. To enable this I tried to add InitOpenCL(NULL) after InitDX<>() call. And I switched OBS to submit RGB frames. Everything worked for me. If you experience crashes please provide exact code (maybe commented out in GitHub). Thanks,

**Comment by Xaymar** [comment]
> I reviewed the dump. It doesn't have a thread that sits inside amfrt64.DLL. There are two threads sitting inside the plug-in (enc-amf.dll, thread IDs = 3332, 17724). What is the reason to believe that the plug-in is blocked by AMF?

The plugin queues up incoming frames, which leads to an increase of memory usage. If it wasn't stuck, it would happily send these off to the AMF component. If it is stuck, it only queues them up, but never dequeues them. This behaviour is not AMF SDK exclusive and also happened with Media SDK - even in the examples.

A normal dump from obs64.exe with my plugin is about 394mb big, a process dump where it got stuck is larger than that. This is easily visible when running OBS in the debugger, as the growth of memory will be quite visible in the Diagnostic View.

> About hick-ups: By default color space conversion in OBS is done on D3D HW queue. It may interfere with other applications. AMF provides you an option to use a different queue: Compute HW queue for the conversion. In this case there is much less interference. To enable this I tried to add InitOpenCL(NULL) after InitDX<>() call. And I switched OBS to submit RGB frames. Everything worked for me.

I will try this on my end too, though I don't think it'll do much since it's not OBS that gets slower (OBS has Realtime priority on my System). If it was OBS, I would not receive 60 frames per second to queue/send of to the submission thread.

> Please separate issues. 

The issues are separate, it's just that you requested the same in both issues.

> Everything worked for me. If you experience crashes please provide exact code (maybe commented out in GitHub).

I'm not sure how to provide more exact code than I already have provided. I could also upload the OBS .json files that will cause the issue if that helps.

**Comment by MikhailAMD** [comment]
I am confused. the plug-in accumulates frames in the queue and memory grows - understood. But what is the reason for not submitting frames? I see three possible variants:
1. AMF did not return from SubmitInput()  - should be visible in the dump
2. AMF returned "INPUT_FULL" - should see tracing and it means that encoder output is not polled fast enough
3. Some synchronization or logical issues in the plug-in

Hick-up: We are talking about GPU priorities. They are not affected by CPU priorities. Compute queue is the way to un-tangle GPU-intensive application (game, browser) and color space conversion. Of course, switching to RGB will increase PCI-E traffic. This problem is going back to the discussion to submit GPU frames to AMF without moving them to system memory.

**Comment by Xaymar** [comment]
The reason is what you described in 1. I'll try to get a dump in which it is stuck inside SubmitInput so it can be seen better.

**Comment by Xaymar** [comment]
I can't seem to get OpenCL working anymore, it continuously fails with the following errors:

```
2016-09-13 21:19:14.543     2F28 [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(1599):OpenCL failed, error = -1:Init::clGetDeviceIDs failed
[AMFTrace::AMFDeviceComputeImpl]    Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(1599):OpenCL failed, error = -1:Init::clGetDeviceIDs failed
2016-09-13 21:19:14.543     2F28 [AMFContextImpl]   Error: ..\..\..\..\..\runtime\src\core\ContextImpl.cpp(317):AMF_ERROR 19 : AMF_OPENCL_FAILED: pDeviceOpenCL->Init(pCommandQueue)
[AMFTrace::AMFContextImpl]    Error: ..\..\..\..\..\runtime\src\core\ContextImpl.cpp(317):AMF_ERROR 19 : AMF_OPENCL_FAILED: pDeviceOpenCL->Init(pCommandQueue)
```

And using InitOpenCLEx with GetComputeFactory fails with AMF_NOT_SUPPORTED. How did you get it to work?

**Comment by MikhailAMD** [comment]
Try to reinstall driver. It should work out of the box. Also try one of AMF samples with OpenCL option: VCEEncoderD3D: in VCEEncoderD3D.bat change -render to  OPENCLDX9

Mikhail

**Comment by Xaymar** [comment]
Same error in the examples. I'll have to try reinstalling the driver then.

**Comment by Xaymar** [comment]
I got the examples work now, seems that the beta drivers were missing OpenCL drivers. DX11 with OpenCL is slower than DX9 with OpenCL there, I hope that this is just an anomaly for the examples.

**Comment by MikhailAMD** [comment]
In general, OCL is slower on introp between DX and OCL (which is needed inside AMF) but it can be masked if you limit performance by 60 fps.

Mikhail

**Comment by Xaymar** [comment]
I have the plugin working with OpenCL and DX11 & 9 now, but it is not useable in any way. FPS is much lower than expected, going as low as 2 encoded frames per second when submitting RGBA frames. What else is needed to get it to encode properly?

**Comment by Xaymar** [comment]
I reverted it to use NV12 frames that are sent from OBS instead now, since that is the only configuration that manages solid 60+ fps on my GPU and CPU.

Also here is a dump where it got stuck in SubmitInput, ThreadId 6708 "enc-amf Input Thread". I hope that it can be fixed.

* xaymar.com: http://cdn.xaymar.com/private/2016/09/amf19_13-09-2016_obs64_stuck_in_SubmitInput.7z
* mega.co.nz:

**Comment by MikhailAMD** [comment]
The dump is OpenCL specific. It should not happen with NV12 submission - the original reported problem.
How to reproduce this?
Is it real hang or you just paused in debugger?
The thread states: 
Input: SubmitInput() calls OpenCL for un-introp DX11 surface back after conversion - need to be sure we stuck there
Output: QueryOutput() waits for SubmitInput to finish - normal

**Comment by Xaymar** [comment]
> The dump is OpenCL specific. It should not happen with NV12 submission - the original reported problem. How to reproduce this?

1. You'll need OBS Studio 0.15.4 and the latest (1.3.0rc3 at the time of writing) plugin version installed.
2. Extract [this archive](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/files/470927/obs-studio.zip) to %appdata%\obs-studio, overwrite where necessary.
3. Start OBS Studio 0.15.4 64-Bit and it should have some sources and similar set up including the full profile that causes the issue to happen.
4. Start recording.

Here are the possible triggers for the issue:

* GPU-heavy tasks in chrome (only happens once): Twitch.Tv streams/decoding, Flash Player
* DirectX 12 and Vulkan applications. Doom and Tomb Raider are currently among the top causes at maximum settings.
* Literally randomly happens during encoding for no reason at all. Nothing out of the ordinary, it just happens. Usually after 30 minutes or more.

> Is it real hang or you just paused in debugger?

It's a real hang. It was stuck there for about ~8 seconds (after Visual Studio finished writing the dump file) before going back to being normal and pushing out encoded frames again. I triggered this one by opening twitch.tv/amd in chrome.

**Comment by Xaymar** [comment]
I just had one of the stuck issues in which it froze the whole system without the OpenCL and DX code enabled. Did not manage to grab a memdump in time since it was near impossible to click the button for it.

**Comment by Xaymar** [comment]
To add on to this one: https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/19#issuecomment-246855174

It is always exactly the same position in which it is stuck - right after ->SubmitInput is evaluated in the submission thread and on the } while (m_Flag_IsStarted); for the querying thread. Seems to be timer resolution related - i've not yet found the exact trigger for this one.

**Comment by MikhailAMD** [comment]
Wait a sec. 
1. If it returned to normal it is not a real hang.
2. if you see in the debugger the code right after SubmitInput() call it can only mean one thing: you have many entries in the queue and the plug-in is trying to submit them in the loop.
3. This is all points to one problem: currently color conversion is done outside of amf so amf is using GPU 3D queue for only one thing: transfer memory from system to GPU. And this may interfere with other applications. I had this problem before with one of game companies. Here is how we resolved this moving the copy to the HW compute queue:
- in Init function:
- Call InitOpenCL() after InitDX<>()
- Acquire AMFCompute interface :m_pContext->GetCompute(AMF_MEMORY_OPENCL, AMFCompute** ppCompute)
- when you convert frame from system memory to AMF:
- Allocate AMF surface as DX9 or DX11 NV12
- surface->Convert(AMF_MEMORY_OPENCL)
- call compute->CopyPlaneFromHost(surface->GetPlaneAt(0),....) //Y plane
- call compute->CopyPlaneFromHost(surface->GetPlaneAt(1),....) // UV plane
- call compute->FinishQueue()
- call surface->Convert(AMF_MEMORY_DX9)
- Submit the frame t othe encoder

- you may want to do all this in the input thread

**Comment by Xaymar** [comment]
These two are not the same technical issue, one is [actually stuck inside SubmitInput](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/19#issuecomment-246852599) (1) and the other is [always returning AMF_INPUT_FULL](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/19#issuecomment-246855174).

The one in which it is stuck in SubmitInput is one that has been reported all across the board - it is an actual full stop of the encoder, no packets are returned or accepted. This stop/hang can last as little as a few seconds or happen over the span of an entire hour after which it will then return packets with 0 byte size. If you want to tell my users that this issue doesn't exist, please make an official blog post then.

The second one is one that freezes the computer for short periods at a time (~0.5 seconds). Input, interaction, rendering, audio, network, etc - all are frozen in time when it happens. I can open a new issue for this one if needed.

> This is all points to one problem: currently color conversion is done outside of amf so amf is using GPU 3D queue for only one thing: transfer memory from system to GPU. And this may interfere with other applications. I had this problem before with one of game companies. Here is how we resolved this moving the copy to the HW compute queue:

I'll have to try this solution. Does CopyPlaneFromHost accept non-aligned memory pages (linesize smaller than minimum padding size)?

**Comment by MikhailAMD** [comment]
for the last question: yes: there is a pitch parameter.

For stuck issues: I am just trying to get as much information as possible. We seem have four cases now. Correct me if I am wrong:
#1 : inside SubmitInput() - the original case was without OpenCL - don't have dump yet
#2 : inside SubmitInput() - OpenCL - the dump was provided 
#3 : encoder return AMF_INPUT_FULL. -  plug-in need to check if it get outputs fast enough
#4 : computer freezes for 0.5 sec and everything stops at this point. - this sounds like display driver is restarting. After this you may have all kind of errors - nothing in D3D will work normally. You should have some AMF error tracing after this.

I will try to get the latest and reproduce this later today. 
Could you please check #3 ? You recently changed the threading logic. The sleep(1) is needed only if you didn't get output buffer. The sleep resolution maybe an issue. Please try to increase it.

**Comment by Xaymar** [comment]
> for the last question: yes: there is a pitch parameter.

Sweet, that makes it a lot easier.

> For stuck issues: I am just trying to get as much information as possible. We seem have four cases now. Correct me if I am wrong:

1. Yup, correct. A user provided a dump file for the original non-OpenCL case [here](https://obsproject.com/forum/threads/amd-advanced-media-framework-encoder-plugin-for-obs-studio.52305/page-11#post-238625).
2. Yep, that is correct.
3. 3 and 4 are the same one, it's just impossible to break all threads since the entire PC is unresponsive and SubmitInput then returns AMF_INPUT_FULL. I've not observed any other D3D applications doing anything out of the ordinary but I'll check if I get tracing output once it happens again in addition of a dump file.

I had changed the threading logic to sleep if GetOutput didn't return AMF_OK (which I think is the correct way to do things right now). Also timer resolution should be at 1ms (Windows Power Usage thing reports that obs64.exe is causing a high timer frequency of 0.5ms).

**Comment by Xaymar** [comment]
> Here is how we resolved this moving the copy to the HW compute queue:
> 
> in Init function:
> Call InitOpenCL() after InitDX<>()
> Acquire AMFCompute interface :m_pContext->GetCompute(AMF_MEMORY_OPENCL, AMFCompute** ppCompute)
> when you convert frame from system memory to AMF:
> Allocate AMF surface as DX9 or DX11 NV12
> surface->Convert(AMF_MEMORY_OPENCL)
> call compute->CopyPlaneFromHost(surface->GetPlaneAt(0),....) //Y plane
> call compute->CopyPlaneFromHost(surface->GetPlaneAt(1),....) // UV plane
> call compute->FinishQueue()
> call surface->Convert(AMF_MEMORY_DX9)
> Submit the frame t othe encoder
> 
> you may want to do all this in the input thread

I currently have the following code:

```
	if (m_AMFCompute == nullptr)
		m_AMFContext->GetCompute(amf::AMF_MEMORY_OPENCL, &m_AMFCompute);

	amf_size l_size0[] = { m_FrameSize.first, m_FrameSize.second, 0 };
	amf_size l_size1[] = { m_FrameSize.first, m_FrameSize.second >> 1, 0 };

	res = m_AMFContext->AllocSurface(memoryTypeToAMF[m_MemoryType], surfaceFormatToAMF[m_SurfaceFormat], m_FrameSize.first, m_FrameSize.second, &pSurface);
	m_AMFCompute->CopyPlaneFromHost(frame->data[0], l_size0, l_size0, frame->linesize[0], pSurface->GetPlaneAt(0), false);
	m_AMFCompute->CopyPlaneFromHost(frame->data[1], l_size1, l_size1, frame->linesize[1], pSurface->GetPlaneAt(1), false);
	m_AMFCompute->FinishQueue();

	// Convert Frame Index to Nanoseconds.
	amf_pts amfPts = (int64_t)(frame->pts * (m_FrameRateReverseDivisor * 10000000.0));
	pSurface->SetPts(amfPts);
	pSurface->SetProperty(L"Frame", frame->pts);
```

I didn't understand why I shouldn't just directly create the surface in OpenCL, since I later convert to OpenCL anyway - is there a performance penalty for creating the surface as OpenCL directly?

**Comment by MikhailAMD** [comment]
The encoder understands DX9 or DX11 surfaces only. If you just submit OCL surface it will do copy through system memory again.
This way you create dx surface, interop it to OCL (just wrapping, no copy), copy data using OCL and compute queue from system memory to video memory, un-interop back to DX, submit to encoder.

Mikhail

**Comment by Xaymar** [comment]
I'm still receiving the following error:

```
2016-09-14 18:34:30.522     42EC [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
[AMFTrace] 2016-09-14 18:34:30.522     42EC [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
2016-09-14 18:34:30.522     42EC [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
[AMFTrace] 2016-09-14 18:34:30.522     42EC [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
2016-09-14 18:34:30.534     42EC [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
[AMFTrace] 2016-09-14 18:34:30.534     42EC [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
2016-09-14 18:34:30.534     42EC [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
[AMFTrace] 2016-09-14 18:34:30.534     42EC [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
```

**Comment by MikhailAMD** [comment]
You missed few operations from the list I sent:

//MM need to force DX allocation – if m_MemoryType is host

    res = m_AMFContext->AllocSurface(memoryTypeToAMF[m_MemoryType], surfaceFormatToAMF[m_SurfaceFormat], m_FrameSize.first, m_FrameSize.second, &pSurface);

//MM need to convert surface to OCL here

    m_AMFCompute->CopyPlaneFromHost(frame->data[0], l_size0, l_size0, frame->linesize[0], pSurface->GetPlaneAt(0), false);

    m_AMFCompute->CopyPlaneFromHost(frame->data[1], l_size1, l_size1, frame->linesize[1], pSurface->GetPlaneAt(1), false);

    m_AMFCompute->FinishQueue();

//MM need to convert surface to DX here
Mikhail

**Comment by Xaymar** [comment]
I've updated my code now, still getting the same error, -30 means CL_INVALID_VALUE if I'm reading the docs right.

```
	amf_size l_size0[] = { m_FrameSize.first, m_FrameSize.second, 0 };
	amf_size l_size1[] = { m_FrameSize.first, m_FrameSize.second >> 1, 0 };

	res = m_AMFContext->AllocSurface(amf::AMF_MEMORY_DX11, surfaceFormatToAMF[m_SurfaceFormat], m_FrameSize.first, m_FrameSize.second, &pSurface);
	pSurface->Convert(amf::AMF_MEMORY_OPENCL);
	m_AMFCompute->CopyPlaneFromHost(&frame->data[0], l_size0, l_size0, frame->linesize[0], pSurface->GetPlaneAt(0), true);
	m_AMFCompute->CopyPlaneFromHost(&frame->data[1], l_size1, l_size1, frame->linesize[1], pSurface->GetPlaneAt(1), true);
	m_AMFCompute->FinishQueue();
	pSurface->Convert(amf::AMF_MEMORY_DX11);
	AMF_LOG_DEBUG("-- Surface Done --");
```

Not sure what else I can change here.

**Comment by MikhailAMD** [comment]
1.      The parameter amf_size origin[3] should be = {0,0,0}; //MM left & top in rect

2.      The parameter amf_size l_size1[] = { m_FrameSize.first>>1, m_FrameSize.second >> 1, 0 };//MM it is in pixels

Mikhail

**Comment by Xaymar** [comment]
Applied the suggested changes now, the error doesn't seem to have changed.

```
[AMFTrace] 2016-09-14 19:15:13.925     280C [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
2016-09-14 19:15:13.941     280C [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
[AMFTrace] 2016-09-14 19:15:13.941     280C [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
2016-09-14 19:15:13.941     280C [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
[AMFTrace] 2016-09-14 19:15:13.941     280C [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
2016-09-14 19:15:13.958     280C [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
[AMFTrace] 2016-09-14 19:15:13.958     280C [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
2016-09-14 19:15:13.958     280C [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
[AMFTrace] 2016-09-14 19:15:13.958     280C [AMFDeviceComputeImpl]   Error: ..\..\..\..\..\runtime\src\core\DeviceComputeImpl.cpp(839):OpenCL failed, error = -30:CopyImageFromHost() clEnqueueWriteImage() failed
```

The errors happen on SubmitInput, so I'll probably have to move surface allocation over to the submission thread as well and only have the queue/dequeue code in ::encode. Unless there is a way to force it to do the copy at that exact moment.

**Comment by Xaymar** [comment]
The trigger for [number 3 from this one]((https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/19#issuecomment-247036027) is apparently to have Twitch open (with a channel/dashboard open), encode and play a game. I have attached a full dump + log file and trace output this time.

Keep in mind that due to the nature of the bug creating a dump file on the position where it halts the system is impossible. I would need to have a virtual machine of which I could dump the full physical state at that exact moment.

In the log file you have to search for "DEBUG BREAK" (without quotes) to see where I paused the process to get a dump file. It shows that while QueryOutput is being called just fine, SubmitInput is slowing down massively - even causing system wide stutters.

Dump & Log:

* xaymar.com: http://cdn.xaymar.com/private/2016/09/amf19_14-09-2016_obs64_AMF_INPUT_FULL.7z
* mega.co.nz: https://mega.nz/#!uRllQRKZ!Xdqc2X6WUTgOSoOZcJL6ywg7BT2fOdGrVdPwKkzLw38

**Comment by MikhailAMD** [comment]
I would like to see whole code.
Mikhail

**Comment by Xaymar** [comment]
It's pretty much at this commit https://github.com/Xaymar/OBS-AMD-Advanced-Media-Framework/tree/f58d288194d20dbcdfa2b9875295e0679fff1c27 with your changes added to the CreateSurfaceFromFrame stuff. I didn't commit anything in between since it didn't work at all.

**Comment by MikhailAMD** [comment]
Why don’t you commit this under some #ifdef so I could see your code and experiment?

Mikhail

**Comment by Xaymar** [comment]
It never occurred to me to use the already existing debug-test branch to push the changes. Here you go: https://github.com/Xaymar/OBS-AMD-Advanced-Media-Framework/tree/debug-test

**Comment by MikhailAMD** [comment]
OK, I fixed few bugs in the branch version that uses OCL - works for me now:
[amd-amf-vce.zip](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/files/475677/amd-amf-vce.zip)

**Comment by MikhailAMD** [comment]
OK, I fixed few bugs in the branch version that uses OCL - works for me now - updated version:
[amd-amf-vce.zip](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/files/475693/amd-amf-vce.zip)

**Comment by Xaymar** [comment]
Can you explain this change to me? Why is both the Y and the UV plane on depth 1?

```
	amf_size l_size0[] = { m_FrameSize.first, m_FrameSize.second, 1 };
	amf_size l_size1[] = { m_FrameSize.first >> 1, m_FrameSize.second >> 1, 1 };
```

**Comment by MikhailAMD** [comment]
this is 3-dimensional. for 2D images depth = 1 - according to OpenCL spec.

**Comment by MikhailAMD** [comment]
I checked the last log -- it is not INPUT_FULL situation (not #3). Compare values for inputFrames and outputFrames in the trace:
20:04:28.089: [AMFTrace] 2016-09-14 20:04:28.017     3924 [AMFEncoderVCE]   Debug: QueryOutput(-1) Repeat: inputFrames=10097, outputFrames=10094
The diffrence is way less then 16 - HW queue size. In fact it is 3-4 which matches b-frame 
numbers.
What you see is a delay in SubmitInput() caused by busy 3D queue - OCL version should partially help.

**Comment by Xaymar** [comment]
Still happens. I will split up the individual issues into their own now so it gets easier to track.

**Comment by Xaymar** [comment]
This Issue is now only dedicated to 1 and 2, since their behavior is really similar in nature. 3 is tracked in #22 now.

**Comment by Xaymar** [comment]
Users have reported that this issue goes away when disabled B-Picture Reference. Which leaves the question if it is due to missing bandwidth or missing data. I will try to get a test case for this that causes it every time.

**Comment by Xaymar** [comment]
No longer able to reproduce this. Might have been fixed by one of the changes in 16.10.*
